### ISSUE TITLE
License being lost w/ Paddle 3.0.25.

### ISSUE THREAD

**Issue by tunabelly** [PaddleHQ/Mac-Framework#166]
@GregularExpressions I have a customer who just emailed saying that the license for TG Pro is being lost every time he restarts (not sure if it's the app or system). Here is his original email:

> TG Pro repeatedly asks me for my license code upon a reboot or startup. 
> After a start, the trial dialog appears and indicates how many days left I have in the trial. 
> Entering the license code and having it validated only works for the current session.
> 
> Macbook Pro Intel i7 2.7GHz 
> MacOS 10.9.x

I replied to him and cc'ed you as well. Please ask whatever you need from the customer so this can be resolved once and for all.

**Comment by GregularExpressions** [comment]
@tunabelly thanks, they've come back to me with the files - we'll investigate.

**Comment by GregularExpressions** [comment]
@tunabelly can you clarify for us when in the SDK lifetime you call `setIsSiteLicensed` please?

**Comment by tunabelly** [comment]
@GregularExpressions it's right before calling startLicensing.

Here's the code:

```
Paddle *paddle = [Paddle sharedInstance];
paddle.delegate = self;
	
[paddle setProductId:<product id>];
[paddle setVendorId:<vendor id>];
[paddle setApiKey:<api key>];
[paddle setIsSiteLicensed:YES];

[paddle startLicensing:productInfo timeTrial:YES withWindow:self.window];
```

**Comment by GregularExpressions** [comment]
@tunabelly Thanks for that, we think there may be a race condition between writing to the two license files (default + site licensing). In the sample files you arranged for us one of the license files was missing the license data present in the other.

Moving `[paddle setIsSiteLicensed:YES];` to before you set the various keys, so after the delegate assignment should help mitigate.

**Comment by tunabelly** [comment]
@GregularExpressions while I appreciate having a potential workaround, just remember this requires us to release yet **another** update to fix this Paddle license issue. This will be I believe the **5th update** required. It has and still is upsetting (I retained from using profanity here) our customers. 

And if the workaround doesn't work for the customers (who are extremely upset it's still not fixed), they'll have to wait for 3.0.28.

**Comment by GregularExpressions** [comment]
@tunabelly yes, I appreciate that you've had to ship an undesirable amount of times to get multiple license fixes to your users and that this is frustrating for both you and them - Tom spoke to me about this.

In short, whilst the symptoms were almost always the same (making it appear as one issue) it was a mixture of licensing, legacy upgrade, and site licensing compound issues. Given the frustration it was causing users it made sense to fix the most prevalent issue and release - minimising the impact over time.

However, we agree it was far from ideal. We have and are continuing to make changes to reduce the likelihood of an issue of such scale in future, these include better logging this end of SDK -> API behaviour and a focus on improved maintainability in v4.

For this particular issue after the SDK update the end users might have to activate once more, to get the site licensing + default licensing files back in sync - but then that should be it.
